### ISSUE TITLE
Freezing in SharedElementTransitionChangeHandler.waitOnAllTransitionNames

### ISSUE THREAD

**Issue by sevar83** [bluelinelabs/Conductor#357]
It's very easy to completely freeze your app while it's waiting for a view with a specific transition name that may never come.

Repro:
1. Have a controller with a row of 5 images in landscape and 3 in portrait. The type of container layout doesn't matter.
2. Click on the 4th or 5th image while in landscape mode.
3. Correct entering shared element transition is displayed.
4. Rotate to portrait.
5. Now it will freeze because the 4th and 5th images are not shown and `SharedElementTransitionChangeHandler` will wait forever.

There are many situations where this problem may happen. **Is there a suggested way to handle the case when a layout in different screen orientations show different sets of images?**

**Comment by EricKuck** [comment]
The suggested way to handle this is to simply not instruct the library to wait on Views that will never appear. I'm not sure what else could really happen in this situation. How would you like to see this handled?

**Comment by sevar83** [comment]
> How would you like to see this handled?

For me it's enough to have a timeout via constructor parameter or a hard-coded reasonable value (say 100ms). If the timeout period has passed the transition to be simply discarded. The shared element transitions usually make sense only for a very short period after the click.

**Comment by EricKuck** [comment]
I'd considered this before, but a time-based solution would be very hard to manage. On older phones with poor layouts 100ms might not be enough to get everything inflated. On newer phones, 100ms is a really long time to lock up the UI and will result in unnecessary dropped frames. I don't think I would want to add anything like this to the library, as developers would be tempted to be overly reliant on the timeout, and therefore end up with jittery apps.

**Comment by sevar83** [comment]
Then I think to pass the screen orientation to the exit change handler. Then on exit if orientations don't match I should either return null transition in `getSharedElementTransition()` or check if the image will be outside of the first controller. Not sure if that can be implemented but I'll try. Thanks for the support!

**Comment by sockeqwe** [comment]
The Problem is that if you use RecyclerView and an item in your recyclerview is the shared element you have to wait until recyclerview and the item in the recyclerview are ready (layout done and ready to draw) and then you can check if the given item that should be animated as shared element  is still there so that shared element transition can run. From my point of view this problem is hard to solve in a generic way that works for all use cases. For example if you additionally have to wait until a bitmap for an image for the shared element item has been loaded and so on. So your porblem is definitely solvable (take a look at  #193 as a starting point) but I'm afraid that you have to write your own `ChangeHandler` to solve this properly for your specific use case.

**Comment by PaulWoitaschek** [comment]
Hmm, I have a vague idea. In activities you can can do something like .postponeEnterTransition() and startPostponedEnterTransition() . 

Maybe there is a way to invert the control and let the controller postpone and signal when it's ready.

**Comment by sockeqwe** [comment]
I, personally, dont like that API neither in Activity, nor in Fragments. I
think it should work fine with ChangeHandlers. You just can't expect that
Conductor has a built-in handler for all kind of share element transitions
out of the box.  In Activity /Fragment  its also up to you to determine by
yourself if the UI widgets are ready and then you call. The only "need"
(not sure if it's really needed) I can thinking of when such a Api like
controller.startPostponedTransition() for confroller could be useful is is
you want to load a bitmap (i.e. picasso) and you want to start animation
after bitmap loaded.

**Comment by PaulWoitaschek** [comment]
The image loading is exactly my use case. It's very hard to get it right from the change handler.

**Comment by sevar83** [comment]
> It's very hard to get it right from the change handler.

@PaulWoitaschek, what problems do you face? I'll try to help.

**Comment by sockeqwe** [comment]
@PaulWoitaschek I feel your pain.
What I typically do is: I pass the Image URL to the ChangeHandler and store it into the Bundle of the change handler. Then the ChangeHandler triggers the image loading and runs the shared element transition once image loaded. 
I also played around with the idea to pass the raw bitmap to the change handler and set the bitmap directly on the shared element ImageView. If bitmap is null (i.e. after process death), I fallback to load the URL. Not sure if I like that idea.

Do you know any better solution?

**Comment by PaulWoitaschek** [comment]
Nope not yet. But is there any way of getting the current change handler from within the controller?
Then you could cast it handle the postponing when necessary.

**Comment by sevar83** [comment]
Hey, what means the tag `hacktoberfest`?

**Comment by The0ldM0nk** [comment]
@sevar83 
Hacktoberfest is [here](https://github.com/blog/2433-celebrate-open-source-this-october-with-hacktoberfest) and will finish [soon](https://hacktoberfest.digitalocean.com/) so get started 🔧
